### PR TITLE
Simplify CI & development experience

### PR DESCRIPTION
**Pull request by vankesteren**
Here is my proposal for a simpler development and CI experience. 

- Only use ruff for linting, which is now in a really good state to replace pydocstyle, flake8, pylint, etc
- Improve CI actions:
  - Scheduled CI on main and development every week
  - Standard CI now runs in parallel (lint/check, test, and doc build) through jobs
  - publish to pypi and dockerhub when a new release is created
- Separate optional dependencies into `test`, `check`, `extra`, `docs`, and `dev` (which contains everything)

Dev process will be: 
- (fork and) clone the repo
- `pip install -e .[dev]`
- create changes
- `ruff check`
- `mypy metasyn`
- `pytest tests`
- `pytest --nbval-lax examples`
- and you're good to commit!

### CODE REVIEW

**Comment by vankesteren** [comment]
I've added pylint rules back in through ruff and removed mentions of pylint

**Comment by vankesteren** [comment]
I'll add more pylint rules next

**Comment by vankesteren** [comment]
Added all pylint rules supported by ruff. Also added pydocstyle rules (and excluded some particularly ignored ones)

**Comment by vankesteren** [comment]
Fixes #282

**Comment by vankesteren** [comment]
After the last commit, I double-checked that the documentation still builds on rtd:

https://metasyn.readthedocs.io/en/update-workflows/

**Review by qubixes — CHANGES_REQUESTED** [review]
Okay for everything except for the formatting changes. Personally, I don't think they are an improvement, but I can be convinced!

**Comment by vankesteren** [comment]
Ah yes, you are right, will fix

**Comment by vankesteren** [comment]
@qubixes are you happy like this?

File: metasyn/varspec.py [hunk 1]
```diff
@@ -64,6 +64,7 @@ def parse(cls, dist_spec: Optional[Union[dict, type[BaseDistribution], BaseDistr
         ------
         TypeError
             If the input has the wrong type and cannot be parsed.
+
```
Comment (qubixes) [inline metasyn/varspec.py:67]: I missed this variation, but this extra line also seems a little bit extraneous? If I look at PEP 257 it doesn't seem to mention anything like that?

Comment (vankesteren) [reply metasyn/varspec.py]: Your're right, good catch, this was also introduced when I hadn't yet configured ruff properly

**Comment by qubixes** [comment]
Okay, approved then, thanks!

**Pull request merged**